### ISSUE TITLE
ABC panels with fm6126A (128x64) require --led-row-addr-type=3 or --led-row-addr-type=5

### ISSUE THREAD

**Issue by marcmerlin** [hzeller/rpi-rgb-led-matrix#823]
This is starting to get annoying...
I get output on them after sending the fm26126A init sequence, but the output is all wrong given the line addressing having changed again. It seems that you only select one of 8 lines, and push 4 times 128 pixels to reach the other 3 rows after filling the first one.
Why, oh why do "they" have to keep changing how to do things? I got those as replacements for a perfectly working ABCDE panel

![IMG_20190525_143255](https://user-images.githubusercontent.com/1369412/58374864-ee830b00-7efa-11e9-8a4d-401715dbf751.jpg)
![IMG_20190525_143240](https://user-images.githubusercontent.com/1369412/58374867-f80c7300-7efa-11e9-9532-de02b828b9a2.jpg)

**Comment by jake653** [comment]
Previous discussions of the quirky  fm6126A"

https://github.com/hzeller/rpi-rgb-led-matrix/issues/746

There is some code and comments regarding that chip, in this driver
https://github.com/2dom/PxMatrix/blob/master/PxMatrix.h

One guy suggests that only init code must be added to solve his FM6126A problem


http://ledpixelart.com/pixelv2board/ declares the FM6126A is not compatible with their setup.

One user post:
"I just ported this file so that it works on my rPi, ran it, and then plugged my Teensy with SmartMatrix code (unmodified), and things just worked.
So this confirms that the only problem is indeed the init necessary with those FM6126A chips"

https://community.pixelmatix.com/t/smartmatrix-doesnt-support-fm6126a-driver-chips/421

**Comment by marcmerlin** [comment]
"One guy suggests that only init code must be added " => that was me :)

Sorry that my message was unclear, the problem is not FM6126A which is solved with the out of band init code (hopefully will get integrated in the lib proper soon), it's the ABC addressing.
@hzeller  can you confirm that ABC addressing is not currently supported by your lib (or any other lib I've seen actually)?  I've only ever seen AB or ABCD or ABCDE

**Comment by mcpgza** [comment]
> Sorry that my message was unclear, the problem is not FM6126A which is solved with the out of band init code (hopefully will get integrated in the lib proper soon), it's the ABC addressing.

I have ABC addressing panel using ICN2037, and I was not able to make it work.

**Comment by hzeller** [comment]
If this is an FM61A but only has three address lines, would this probably be working with (the new flag) `--led-panel-type=fm6126` and one of the --led-multiplexing options (because outdoor panels do different multiplexing).

**Comment by mcpgza** [comment]
My 64x128 panel is not fm6126, but ICN2037, and it has only ABC lines:

![20190905_222416_2](https://user-images.githubusercontent.com/50031926/64378517-2a1a8e80-d02d-11e9-8374-c5f739dab9db.jpg)

and this is how the demo with -D 3 looks like:
![20190831_233101](https://user-images.githubusercontent.com/50031926/64378699-8a113500-d02d-11e9-9fa6-4522130c6f82.jpg)

It is written on the panel: P2-128X64-32S-V4.0, LED1515, 2018/06/16

(I have other panels with HUB75E with ICN2037, and those works fine)

**Comment by hzeller** [comment]
ABC sounds like a different led multiplexing. Did you try through the available `--led-multuplexing` options?

**Comment by mcpgza** [comment]
Yes, I tried them all, none of them worked. 

This is how demo 0 looks like:

sudo ./demo  --led-cols=128 --led-rows=64 --led-slowdown-gpio=5   -D 0

![ezgif-3-04a03bf5624c](https://user-images.githubusercontent.com/50031926/64515899-c07cd780-d2ed-11e9-9d41-bf8fa0ed2d44.gif)

**Comment by hzeller** [comment]
ABC could also means some different row address type; there are a few implemented, but they don't do three address lines. It might be worthwhile doing some reverse engineering, then implement this.

It looks a bit like the AB row address type but maybe split into top and bottom half ? So I would try with `--led-row-addr-type=1` and see if that at least gives some partial usable output. Then we can go from there.

**Comment by themultiplexer** [comment]
I also have one of these ABC-panels and have the same issue as @5onix, @marcmerlin and @mcpgza. 
I'd love to see this awesome library work with my panel. I would like to help but I don't know where to start 😅
Maybe I can help after I heard Computergraphics this semester.

**Comment by hzeller** [comment]
Good starting point is to get hold of a datasheet

**Comment by gustavoalara** [comment]
Hi, 

I have the same case as @mcpgza. I received two days ago two 128x64 led-matrix with the same ABC addresses and same information written on the panel (P2-128X64-32S-V4.0, LED1515, 2018/06/16). I asked the chinese store from Aliexpress where I bought the panel for some schematics or something useful but they said that their have no information about it

**Comment by gustavoalara** [comment]
Hi again,

According to the adafruit site (https://learn.adafruit.com/32x16-32x32-rgb-led-matrix/new-wiring), the ABC addresses is the pin arrangement used by 32x16 panels
![led_matrix_socket1](https://user-images.githubusercontent.com/7902641/65247986-bb8c0500-daf1-11e9-99a5-840fff689689.png)

I don't know if this could be useful to someone. Maybe this panel needs a custom internal mapper...

**Comment by bk313** [comment]
I also have this 128x64 panel with the ICN2037 and the ABC lines. No luck getting it to work, I get the same image as @mcpgza posted above. Would like to help any way I can.

**Comment by rowanG077** [comment]
I have this panel as well. It seems a completely different addressing scheme then before.

**Comment by mcpgza** [comment]
It seems it requres a new RowAddressSetter, and I would like to wrtire it, but unfortunately I dont understand how it works, dear @hzeller, can you provide a short descrption how it works, what it does. Thanks.

**Comment by rowanG077** [comment]
I have looked at the code for the AB address setter. The way it seems to work is that `A` is a clock
signal and `B` is the data. You clock on the `A` signal and when you have reached the `addressableRows - rowToSelect` clock you set `B` to low for that clock cycle. As an example
if you have a panel with 8 addressable rows you and you want to select the third row you would first
clock the `A` line for 5(8 - 3) cycles while keeping `B` high and then you would clock the `A` line for a single cycle and have `B` low. And the clock `A` for the remaining two cycles while keeping `B` high again. Below is an illustration:

![abAddress](https://user-images.githubusercontent.com/7439756/65316619-9b645080-db9a-11e9-8bb8-fea3c9e2852d.png)

Here is the relevant [code](https://github.com/hzeller/rpi-rgb-led-matrix/blob/7160a8ade3dcecabe25ce0de73232b365f7587be/lib/framebuffer.cc#L127).
```
  virtual void SetRowAddress(GPIO *io, int row) {
    if (row == last_row_) return;
    for (int activate = 0; activate < double_rows_; ++activate) {
      io->ClearBits(clock_);
      if (activate == double_rows_ - 1 - row) {
        io->ClearBits(data_);
      } else {
        io->SetBits(data_);
      }
      io->SetBits(clock_);
    }
    io->ClearBits(clock_);
    io->SetBits(clock_);
    last_row_ = row;
  }
```

What I don't understand is  that extra clock low and clock high after the loop. Anyway this is just an analysis for the AB line setter and I don't have such a panel. The ABC addressing may be similar to this or completely different.

**Comment by gustavoalara** [comment]
According to Adafruit for the 32x16 Led matrix,
 
A is the lowest bit of row address
B is the second lowest bit of row address
C is the highest bit of row address
LE is Latch of row
CLK is Clock
EN is  Enable
R1 is Red data 1
R2 is Red data 2
G1 is Green data 1
G2 is Green data 2
B1 is Blue data 1
B2 is Blue data 2
GND is Grounding (and 4 GND must be connected)

I think ours panels is the same pinout but with a 1:32 scan multiplexing... Anyway, I've contacted the manufacturer of my led matrix (Coreman.cc). Let's see if I'm lucky and they answer me with the information I've requested

**Comment by rowanG077** [comment]
The addressing of the 32x16 Adafruit panels are the same as the usual A,B,C,D addressing but it's just using the A,B and C lines directly since it just needs to address 8 rows. This addressing on the panels here is completely different.

**Comment by gustavoalara** [comment]
Hi,

I found this forum with some information that can be useful https://community.pixelmatix.com/t/led-module-p2-128-64-1-32-driver/414

**Comment by rowanG077** [comment]
That could be it. I hadn't found that yet even though I searched that forum. Maybe I can take a look this weekend if this will work.

**Comment by mcpgza** [comment]
I made a huge progress in refactor. the panel does not contain inverters, so it use "positive logic", and it seems tha panel has AB type addressing, but B is not used, C used instead B.

sudo ./demo --led-cols=128 --led-rows=64 --led-slowdown-gpio=5 --led-row-addr-type=1 -D 0
![ezgif-4-d5961fdf881f](https://user-images.githubusercontent.com/50031926/65410965-ed4fe500-ddeb-11e9-9419-88d35d9b79c1.gif)

This is the relevant code: 
`class ShiftRegisterRowAddressSetter : public RowAddressSetter {
public:
  ShiftRegisterRowAddressSetter(int double_rows, const HardwareMapping &h)
    : double_rows_(double_rows),
      row_mask_(h.a | h.c),
      clock_(h.a),
      data_(h.c),
      last_row_(-1) {
  }
  virtual gpio_bits_t need_bits() const { return row_mask_; }

  virtual void SetRowAddress(GPIO *io, int row) {
    if (row == last_row_) return;
    for (int activate = 0; activate < double_rows_; ++activate) {
      io->SetBits(clock_);
      if (activate == double_rows_ - 1 - row) {
        io->SetBits(data_);
      } else {
        io->ClearBits(data_);
      }
      io->ClearBits(clock_);
    }
    io->SetBits(clock_);
    io->ClearBits(clock_);
    last_row_ = row;
  }

private:
  const int double_rows_;
  const gpio_bits_t row_mask_;
  const gpio_bits_t clock_;
  const gpio_bits_t data_;
  int last_row_;
};
`
The only problem is a missing line at the end of one part.

**Comment by bk313** [comment]
Great work @mcpgza !

**Comment by hzeller** [comment]
Thanks @mcpgza ! I have added your code to the master branch now and it can be selected with `--led-row-addr-type=3` so that it is easy to try for everyone.

This then allows others who also have the panel to help figure out the remaining issue with the missing line.

**Comment by gustavoalara** [comment]
Hi guys,

I'm trying the new option for our 128x64 ABC modules. Apart of the black line, I see two things:

1. some LEDs with incorrect colours near the right and left laterals
2. If I try to connect 2 128x64 modules in chain, the modules don't show anything and the application freezes

I'll upload some images of these things

**Comment by gustavoalara** [comment]
Forget the second issue... I incremented the nanoseconds from 300 to 450 and now the chain works fine. Only the of issues with some LEDs:


![IMG_20190925_204323](https://user-images.githubusercontent.com/7902641/65630168-27300f80-dfd5-11e9-83da-51605ced8927.jpg)

**Comment by 5onix** [comment]
For your 1st problem, have you tried the demo #4  - Pulsing color or #11 - Brightness pulse generator ?
Those demos made me realize that my panel had a few bad subpixels on the edges.
The edges are kinda fragile. I had to remove a few leds which weren't working anymore after some accidental bumps they were falling apart.

**Comment by gustavoalara** [comment]
Hi @5onix I've tried with demos and images (as you can see in my previous message) and the result is the bad subpixels on the edges. My second matrix module is failing and the low quarter of the module has no light (I thing something has broken)

Following a picture of the two modules working together, (the left one is the broken module). I have 3 LEDs broken too (in the right corner of the right module)
![IMG_20190925_223704](https://user-images.githubusercontent.com/7902641/65637900-4e8ed880-dfe5-11e9-9d6a-446825d88011.jpg)

**Comment by 5onix** [comment]
Too bad that one of your module has an issue. But overall the result looks very promising.
Here's 2 photos of mine which show broken and malfunctioning pixels.
![shg-led-matrix](https://user-images.githubusercontent.com/12273664/65638621-c6a9ce00-dfe6-11e9-9743-ca97c44ce851.jpg)
![removed-and-bad-pixels](https://user-images.githubusercontent.com/12273664/65638622-c6a9ce00-dfe6-11e9-9345-e2f322bc0407.jpg)

**Comment by marcmerlin** [comment]
speaking of P2 panels and pixels that fall off, have you found some kind of glue spray that you can layer on top of them to glue them all in place so that they don't get lost? (and without affecting display quality and/or the space between panels if you put them adjacent to one another)

**Comment by gustavoalara** [comment]
Good point @marcmerlin, this panels look very fragile and is very easy to lost pixels. (I found 2 of the pixels, but I don't know how to glue them to the panel again)
In the other hand, I'm arguing with the seller because I think the panel arrived broken

**Comment by marcmerlin** [comment]
@gustavoalara you need both. I also had a panel that was not protected correctly in shipping and arrived with some detached pixels, but trust me when I tell you that you will lose more unless you encase this in some way, or never touch/move it, so you might as well practice with the damaged panel :)

**Comment by 5onix** [comment]
Epoxy resin maybe ?

**Comment by gustavoalara** [comment]
Then, you thing the LEDs with incorrect colours are damaged too? Well, maybe if I'm lucky the seller will send me another panel and I can practice with the damaged panel as @marcmerlin said...

**Comment by 5onix** [comment]
I know it's a little off topic but here's a little demo of the panel playing a video with VLC & sound enabled.
[https://gofile.io/?c=F2Mp86](url)

**Comment by mcpgza** [comment]
Ok, here is the fully functional SetRowAddress (no missing line), dear @hzeller please update the code

```
 virtual void SetRowAddress(GPIO *io, int row) {

    for (int activate = 0; activate < double_rows_; ++activate) {
      io->ClearBits(clock_);
      if (activate == double_rows_ - 1 - row) {
        io->SetBits(data_);
      } else {
        io->ClearBits(data_);
      }
      io->SetBits(clock_);
    }
    io->SetBits(clock_);
    io->ClearBits(clock_);
    last_row_ = row;
  }

```

I dont know why is it works, but works.

**Comment by 5onix** [comment]
Awesome progress made here, thank you @mcpgza !

**Comment by gustavoalara** [comment]
Hi!

The change proposed by @mcpgza works like a charm! My only problem now is the dead and malfunctioning pixels
![IMG_20190927_184752](https://user-images.githubusercontent.com/7902641/65786792-b2d1a980-e157-11e9-8f73-8168361e8806.jpg)

![ezgif com-resize](https://user-images.githubusercontent.com/7902641/65787092-7a7e9b00-e158-11e9-8625-284ba11fce25.gif)

**Comment by marcmerlin** [comment]
For dead pixels, in the future, I was told to consider this
https://www.amazon.com/gp/product/B0009GXMJG
not sure how well it works

**Comment by embedded-creations** [comment]
@marcmerlin If you need the LED to be functional, I think you'd need to hot air rework the connections, or replace the LED:

https://twitter.com/GregDavill/status/1172850535403114496

**Comment by marcmerlin** [comment]
@embedded-creations nice. I wasn't the one who lost pixels on those panels (but I lost some on another panel, that were already detached when the panel arrived in the mail), but while I'm impressed by the work done to fix them, that's a bit beyond my equipment/skill level for the moment.
I'll focus on trying to find the right glue spray to put on top of the entire panels when I receive them so that hopefully I don't lose pixels to start with.
My immediate problem with my wearable panels is that because they are flexible, even the slight flex they receive eventually messes solder joints of the chips in the back, making me lose entire columns or rows depending on which pad gets disconnected (not in a visible way). Thankfully it's usually only one color component, so I don't lose the entire pixel, but it's still annoying.
This is out of topic for this bug, so I opened a new discussion here https://community.pixelmatix.com/t/rgb-panels-and-wear-issues-losing-pixels-and-losing-solder-joints-in-the-back-on-flexible-panels/530

**Comment by hzeller** [comment]
Now it is :)
Thanks @mcpgza !

**Comment by marcmerlin** [comment]
@mcpgza @hzeller I've only been able to try my ABC panels (really AC from what you said) now.
It works great with --led-row-addr-type=3 --led-panel-type=FM6126A
BTW, @hzeller  I can confirm --led-panel-type=FM6126A works with the active parallel board, both my panels were initialized perfectly.

**Comment by whiteshavedchocolate** [comment]
Guys, I am pretty new to all of this. However, I have the datasheet for the ICN2037 in English. Would that be of interest/help to anybody?

It looks like the main feature it adds over a standard LED shift register/column driver is that it has "dual latch for higher refresh rate". Maybe that can confirm how @mcpgza got the ABC LED panel to light up properly.

**Comment by marcmerlin** [comment]
@mcpgza @bk313 @gustavoalara @5onix can I ask you to have a look at https://github.com/hzeller/rpi-rgb-led-matrix/issues/985 (I'm somehow not allowed to @ you in that issue).
Long story short, ABC works fine, but it's slower than ABCDE, which it shouldn't be.

**Comment by marcmerlin** [comment]
@whiteshavedchocolate this bug is about ABCDE vs ABC. Using the wrong led-row-addr-type still lights up fine, but displays lines in the wrong places.
ICN2037 vs FM6126A is a shift register issue, and not related to  led-row-addr-type
What is the exactly problem you have and you're trying to fix?

**Comment by whiteshavedchocolate** [comment]
I don't currently have a problem. I am learning and interested in hardware
development around these Chinese LED panels. I am willing to help where I
can, but my skills are in hardware and very limited in software/firmware.

**Comment by marcmerlin** [comment]
@whiteshavedchocolate ok, then there is no problem at the moment, ICN2037 seems to work fine as is (unless I missed something)

**Comment by tdaede** [comment]
FWIW the current code implements 33 clocks into the shift register (an extra clock at the end), but I found that 32 clocks is correct for my FPGA reimplementation. I'm wondering if the last clock is fast enough that it's getting missed. (the last clock actually looks like a leftover from the old version with inverted clock polarity).

**Comment by marcmerlin** [comment]
@mcpgza  @bk313  @gustavoalara  @5onix can I ask you where you bought your panels from, and what you paid for them?
My current supplier is offering them at $50 per panel (shipped from china), and unfortunately he says he can only sell ABC, as ABCDE doesn't seem to be available anymore. 
I'm a bit bummed about having to use ABC panels with a lower refresh rate, but if that's all what is available today, and $50/panel is about the right price, so be it, I'll go ahead and get those
(I need 8, so making sure before I plop down $400 on this)

**Comment by mcpgza** [comment]
@marcmerlin I bought theese: https://www.aliexpress.com/item/32998082589.html
ABC panels, not perfect panels, some of them seem used, some have missing leds on the edge, but generally i'm happy with them for this price.  
But why do you think that ABC panels are slower then ABCDE?

**Comment by marcmerlin** [comment]
Thanks @mcpgza . I'll probably pay a bit more for pristine panels, I don't want missing pixels for my setup :)
https://github.com/hzeller/rpi-rgb-led-matrix/issues/985 explains why ABC is slower than ABCDE. In a nutshell, ABC panels are ABCDE panels with 2 address lines missing, so it takes longer to send the 5 bits of addresses over 3 bits.
you can benchmark this yourself with
~/rpi-rgb-led-matrix/examples-api-use/demo --led-rows=64 --led-cols=128 --led-chain=1 --led-show-refresh --led-parallel=3   --led-pwm-dither-bits=1 --led-pwm-bits=7 --led-panel-type=FM6126A  --led-row-addr-type=0 -D0  and then use led-row-addr-type=3
You can do the speed test even without the right panels and look at the Hz value you get.

**Comment by arash1221** [comment]
Hello @hzeller , First of thanks for your awesome code, please anyone can help me with my problem!!!
My problem is when I'm add another panel to the output of the first panel the screen shows me different output but alone with just 1 led panel everything is okay! and also I using the same led panel!! panel IC is ------> FM6126A_ 64*128
I tried these command flags also not work
 --led-chain=2  && --led-chain=1 
I also set col to 256 and also dosen't work!!!,And also I'm not using any hat, direct to raspberry pi GPIO! Everything is ok when its just only 1 panel no more chain!!!!!

Please Somebody Help!
Appreciate!

**Comment by arash1221** [comment]
First Image Only 1 led Panel

![Untitled-2](https://user-images.githubusercontent.com/20780081/80252580-1ac67f00-86aa-11ea-9bab-963c3bca674c.jpg)

Second Image two chain panel !!!

![Untitled-3](https://user-images.githubusercontent.com/20780081/80252717-6aa54600-86aa-11ea-99a5-4411ab24926d.jpg)

**Comment by mcpgza** [comment]
> First Image Only 1 led Panel

> 
> Second Image two chain panel !!!
> 

Yeah, tipical, you need a level shifter, you need a hat.

**Comment by arash1221** [comment]
No way to use hat??

**Comment by arash1221** [comment]
@mcpgza thanks for the answer, can you please guide me how to control any pixel of led panel,
very simple code, I just want to control each pixel on my own, I tried a lot with nodejs but no result I get!
thanks for your help.

**Comment by hzeller** [comment]
@arash1221 Please don't hijack an existing issue with an unrelated problem; open a new issue.

To your problem: try to gpio slowdown your output (`--led-slowdown-gpio`)  depending on the internal wiring you're increasing the load on the gpio pins, which will make them switch  slower. 

Also, you _really_ want a level shifter, as not only will be the digital voltage correct (and right now you're just essentially playing jeopardy if it might work), but also, you can't really drive long lines with the GPIO pins directly.

**Comment by arash1221** [comment]
Thanks @hzeller  I recall it.

**Comment by xsrf** [comment]
> Thanks @mcpgza . I'll probably pay a bit more for pristine panels, I don't want missing pixels for my setup :)
> #985 explains why ABC is slower than ABCDE. In a nutshell, ABC panels are ABCDE panels with 2 address lines missing, so it takes longer to send the 5 bits of addresses over 3 bits.
> you can benchmark this yourself with
> ~/rpi-rgb-led-matrix/examples-api-use/demo --led-rows=64 --led-cols=128 --led-chain=1 --led-show-refresh --led-parallel=3 --led-pwm-dither-bits=1 --led-pwm-bits=7 --led-panel-type=FM6126A --led-row-addr-type=0 -D0 and then use led-row-addr-type=3
> You can do the speed test even without the right panels and look at the Hz value you get.

Hey, I've seen your discussion about ABC being slower than ABCDE. This is because the code is not using the Shift-Register that is connected to the row selection as it is meant to be used (I guess).

To select one out of N rows, you are shifting out N-1 zeroes and one one out to the display. Of course this needs more than N clock cycles which is way longer on 1/16 or even 1/32 panels than just setting 5 GPIOs at once.

However: This is actually not necessary if you scan your rows incrementally, which is usually what is done. In this case you actually do only have to shift out one bit for every cycle. You shift out 1 for row 0 and 0 otherwise.

Here is the code I contributed to PxMatrix few days ago:
https://github.com/2dom/PxMatrix/blob/178054b3021892cb9185c3a92051ba2b2afc8977/PxMatrix.h#L1012

If you don't require random row access, it boils down to:

```
      // Just shift the row mux by one for incremental access
      digitalWrite(_C_PIN, (value==0) ); // Shift out 1 for line 0, 0 otherwise
      digitalWrite(_A_PIN, HIGH); // Clock out this bit
      digitalWrite(_A_PIN, LOW);
```

Hope this may also help you :)

**Comment by hzeller** [comment]
Nice, thanks xsrf

**Comment by hzeller** [comment]
Nice, thanks @xsrf .
Could someone who has such a panel (I Ithink I don't have any if these in mystash) implement and test that in the rpi-rgb-led-matrix library and send me a pull request (e.g. @marcmerlin :)
I could implement it blindly, but it is better with a board to test that it actually works :)

**Comment by mcpgza** [comment]
@hzeller please implement it, i will test and fix it if necessary. I dont know how to translate that code to your library api.

**Comment by marcmerlin** [comment]
So, I do have ABC panels somewhere in my garage, but I'm not using them, as my main array is ABCDE, so honestly my enticement to spend time on them for a small improvement that i won't see because I don't use them, is kind of limited :)
That said, I'm happy to confrim that new code works on them if that helps.

**Comment by javiervalero31** [comment]
Hi. I read your comments to try to solve my problem but could not solve it. 
I have 4 LED arrays with 32x16 size. 
I want the image or Demos to be centered in my panels but I have not been able to configure the parameters to achieve this.


Each matrixs connected in a only matrix of 64x32 size...
![IMG_20200622_134157](https://user-images.githubusercontent.com/42529408/85319762-04615780-b490-11ea-96bb-cd5feaf7cd6a.jpg)

The closest thing to what I want (DEMO 0)....
![IMG_20200622_141029](https://user-images.githubusercontent.com/42529408/85321218-55724b00-b492-11ea-806b-c33f5155d932.jpg)

Any idea?

**Comment by marcmerlin** [comment]
I believe those panels are not ABC, so this is not the place to ask for him on those panels.
Unfortunately there isn't a support group yet for this library, @hzeller  was consideirng options, otherwise for now try to file a new issue with all the details on your panels, how they are wired, what command line options you used to run the demo, etc.

**Comment by marcmerlin** [comment]
@javiervalero31  actually there is a support group now, I just created one :) 
https://rpi-rgb-led-matrix.discourse.group/

**Comment by Sl-Alex** [comment]
@xsrf I implemented it in my code and after that I found your message.
> Here is the code I contributed to PxMatrix few days ago: https://github.com/2dom/PxMatrix/blob/178054b3021892cb9185c3a92051ba2b2afc8977/PxMatrix.h#L1012
> 
> If you don't require random row access, it boils down to:
> 
> ```
>       // Just shift the row mux by one for incremental access
>       digitalWrite(_C_PIN, (value==0) ); // Shift out 1 for line 0, 0 otherwise
>       digitalWrite(_A_PIN, HIGH); // Clock out this bit
>       digitalWrite(_A_PIN, LOW);
> ```
> 
> Hope this may also help you :)

If you implement it like this then it also fixes the problem with the black line in the middle of the module (which was reported many times in this issue).
I also have the panel with ABC row addressing and I found something interesting today. Just look at ICND2019 datasheet. The control sequence is exactly the same as you proposed. Also it seems that `B` line could be used in order to configure some extended register (haven't tried it yet). Here is a line from the datasheet:
`For control card, SDI is the C of 3-8 decoder, DCK is the A of 3-8 decoder, RCK is the B of 3-8 decoder`

**Comment by marcmerlin** [comment]
this has been fixed for a while, closing

**Comment by marcmerlin** [comment]
@mcpgza @Sl-Alex @xsrf @javiervalero31 , since apparently these ABC panels still exist and are still being shipped, could someone re-confirm my understanding to make sure I'm correct:
Is there any advantage to these panels whatsoever (maybe brighter?) or they are really just the same as ABCDE but with slower addressing to save a few pins and maybe a few cents of construction?

**Comment by rowanG077** [comment]
The big advantage is you can support an arbitrary amount of rows. with ABCDE you are limited to 32 rows.  Although it is slower it's not like it's big. A single cycle extra every row you are clocking out.

**Comment by marcmerlin** [comment]
@rowanG077 let me rephrase to make sure I understand.
You are saying with ABC you can support let's say 128 rows instead of 64 rows, but if you do, now you are refreshing one row out of 64 instead of one row out of 32, so the refresh bar problem is going to be even worse unless you can push updates at very high speed.
So if we're back at 64 rows in 1/32 updates, is ABC better in any way than ABCDE outside of saving 2 GPIOs on the rPi?

**Comment by rowanG077** [comment]
with a good FPGA implementation 128 rows wouldn't really be a problem. Some LED panel usage requires only 1-bit depth for simple signage applications. There you could even imagine 256 rows or 512 rows.

Personally I think a uniform interface is much better then just add a pin everytime. It makes connectors and clients much more expensive. But besides that there is no advantage I'm aware of,

**Comment by marcmerlin** [comment]
Thanks @rowanG077  that's a reasonable answer, In my easier tests ABC seemed slower for ABCDE, but maybe it was just the current implementation. Your point that ABC allows adding even more rows without adding more pins, is indeed valid.

**Comment by marcmerlin** [comment]
I just bought from P1.875 flex panels, they are quite nice outside of the lack of output port (so you you can only put one panel per chain):
![image](https://github.com/user-attachments/assets/1653cd00-2605-4423-9e9e-ea3bee68f889)
They are indeed ABC panels which is not ideal, but they work with 
--led-row-addr-type=3
however, they also needed --led-slowdown-gpio=4 for the display not to be garbled, and that was on a rPi0 2l which is equivalent in speed to an rPi3, so with an rPi4, they may need an even higher lowdown

**Comment by marcmerlin** [comment]
I forgot to give the link for the rpi0 2w vs rpi3 speed tests: https://rpi-rgb-led-matrix.discourse.group/t/rpi0-2w-speed-compared-to-rpi3a/913  (basically they are just as fast)

**Comment by Jan1503** [comment]
Hey @marcmerlin could you share the (aliexpress?) link for these panels?
I'm in the process of ordering some new P2 128x64 panels, but the seller tells me, they use FM6126A chips but with ABCDE address lines, which sounds a bit confusing.
These P1.875 panels look cool and maybe it's possible to find the output pins of the chip on the pcb to allow chaining?!
Thanks!

**Comment by baicker** [comment]
> I just bought from P1.875 flex panels, they are quite nice outside of the lack of output port (so you you can only put one panel per chain): ![image](https://github.com/user-attachments/assets/1653cd00-2605-4423-9e9e-ea3bee68f889) They are indeed ABC panels which is not ideal, but they work with --led-row-addr-type=3 however, they also needed --led-slowdown-gpio=4 for the display not to be garbled, and that was on a rPi0 2w which is equivalent in speed to an rPi3, so with an rPi4, they may need an even higher lowdown

@marcmerlin could you share the link of this P1.87 128x64 pannel ? What type of chipes are?

**Comment by marcmerlin** [comment]
Just a heads up, with at least one set of panels, I need slowdown of 8, which is both surprising and sad :(
https://github.com/hzeller/rpi-rgb-led-matrix/issues/1773

**Comment by marcmerlin** [comment]
@mcpgza @Sl-Alex @xsrf @javiervalero31 @rowanG077 . Github does not seem to allow me to tag you on https://github.com/hzeller/rpi-rgb-led-matrix/issues/1774 , but if you can please sub to the new bug. ABC code does work but it seems to have a timing/speed issue that requires a big slowdowns to avoid garbage output

**Comment by marcmerlin** [comment]
@baicker  I missed your ping here, sorry. As answered on the support board, those panels came from 
https://www.aliexpress.us/item/2255801134833289.html  
Note that only the P1.875 are ABC and do not have an output port.  The P2 are ABCDE and have an output port.
Then please look at https://github.com/hzeller/rpi-rgb-led-matrix/issues/1774 which also affects the panel I just mentioned when using dietpi6 , I had to use a slowdown of 4, but if I go back to raspbian 4, I expect the slowdown will go back to 2 or 1.

Give that slowdowns of up to 8 have been needed with another set of P2 ABC panels, it's fair to re-open this bug until the sub-bug, is closed.

**Comment by marcmerlin** [comment]
for people using A(B)C panels with this row driver, what row chips do you have?
I checked my old panels from 2019, and their row selector seems to be a 74HC245C
This one worked without needing to add delays.

**Comment by marcmerlin** [comment]
Please see https://github.com/hzeller/rpi-rgb-led-matrix/issues/1774
and https://github.com/hzeller/rpi-rgb-led-matrix/pull/1786
If you have recent ABC panels, although shift registers should all work the same, newer panels use new SR chips that do not work well with led-row-addr-type=3 (by needing a huge slowdown) and work much better with led-row-addr-type=3

**Comment by mbuvi-dev** [comment]
![Image](https://github.com/user-attachments/assets/2ff02bce-774a-47d1-a3ca-fb3120ef5902)
with this new board I can't get it working with esp32D 
hahaha nobie

**Comment by mbuvi-dev** [comment]
Do i need a card ?? or can they just work esp and my LED.

**Comment by marcmerlin** [comment]
> Do i need a card ?? or can they just work esp and my LED.

You are looking at the wrong project for ESP32, check out https://github.com/mrcodetastic/ESP32-HUB75-MatrixPanel-DMA
but I don't think it supports ABC panels, however this lib should:
https://github.com/2dom/PxMatrix
